pkg/storegateway/bucket_index_reader.go

```diff
@@ -625,49 +632,74 @@ func (r *bucketIndexReader) preloadSeries(ctx context.Context, ids []storage.Ser
 
-	stats.update(func(stats *queryStats) {
-		stats.seriesFetchCount++
-		stats.seriesFetched += len(ids)
-		stats.seriesFetchDurationSum += time.Since(begin)
-		stats.seriesFetchedSizeSum += int(end - start)
-	})
+	offsetReader := seriesOffsetReaders.Get().(*offsetTrackingReader)
+	defer seriesOffsetReaders.Put(offsetReader)
 
-	for i, id := range ids {
-		c := b[uint64(id)-start:]
+	offsetReader.Reset(start, reader)
+	defer offsetReader.Reset(0, nil) // don't retain the underlying reader
+
+	bytesPool := pool.NewSlabPool[byte](seriesBytesSlicePool, seriesBytesSlabSize)
+	// We never release the pool and let the GC collect it in order to avoid a rance condition with the async sets of the cache.
 
-		l, n := binary.Uvarint(c)
-		if n < 1 {
-			return errors.New("reading series length failed")
+	for i, id := range ids {
+		// We iterate the series in order assuming they are sorted.
+		err := offsetReader.SkipTo(uint64(id))
+		if err != nil {
+			return err
 		}
-		if len(c) < n+int(l) {
+		seriesSize, err := binary.ReadUvarint(offsetReader)
+		if err != nil {
+			return err
+		}
+		seriesBytes := bytesPool.Get(int(seriesSize))
+		n, err := io.ReadFull(offsetReader, seriesBytes)
+		if errors.Is(err, io.ErrUnexpectedEOF) {
 			if i == 0 && refetch {
-				return errors.Errorf("invalid remaining size, even after refetch, remaining: %d, expected %d", len(c), n+int(l))
+				return errors.Errorf("invalid remaining size, even after refetch, read %d, expected %d", n, seriesSize)
 			}
 
+			level.Warn(r.block.logger).Log("msg", "series size exceeded expected size; refetching", "series_id", id, "series_length", seriesSize, "max_series_size", maxSeriesSize)
 			// Inefficient, but should be rare.
-			r.block.metrics.seriesRefetches.Inc()
-			level.Warn(r.block.logger).Log("msg", "series size exceeded expected size; refetching", "id", id, "series length", n+int(l), "maxSeriesSize", maxSeriesSize)
-
 			// Fetch plus to get the size of next one if exists.
-			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+uint64(n+int(l)+1), loaded, stats)
+			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+binary.MaxVarintLen64+seriesSize+1, loaded, stats)
+		} else if err != nil {
+			return errors.Wrapf(err, "fetching series %d from block %s", id, r.block.meta.ULID)
 		}
-		c = c[n : n+int(l)]
-
-		loaded.addSeries(id, c)
+		loaded.addSeries(id, seriesBytes)
 
-		r.block.indexCache.StoreSeriesForRef(r.block.userID, r.block.meta.ULID, id, c)
+		r.block.indexCache.StoreSeriesForRef(r.block.userID, r.block.meta.ULID, id, seriesBytes)
 	}
 	return nil
 }
 
+func (r *bucketIndexReader) recordLoadSeriesStats(stats *safeQueryStats, refetch bool, numSeries int, start, end uint64, loadStartTime time.Time) {
+	stats.update(func(stats *queryStats) {
+		if refetch {
+			stats.seriesRefetches++
+			// We don't track refetches in a lot of detail because they are rare
+			return
```

Comment:
I'm not sure about this. Why not tracking the other stats for refetches too?

Reply:
I think recording time spent fetching is incorrect if we record it twice, but the rest make sense (fetched series & fetched series size)